core/tools/bin_size/src/bin_size/api.py

```diff
@@ -0,0 +1,226 @@
+    data_size: int = 0  # how much of the symbol size is data
+
+    # Added optionally by `BuildDefinitionLoader`
+    build_definition: str = ""  # definition which goes to the build
+
+    # Added optionally by `BinarySize.aggregate`
+    number_of_symbols: int = 0  # how many symbols are aggregated under one
+
+    # Added optionally by `RowHandler.add_source_definition`
+    # In case of C files, this definition is the same as the build definition
+    # In case of micropython, build definition points to `build/firmware/frozen_mpy.c`,
+    # but source definition will be somewhere in src/apps/
+    # Rust files do not even have the build definition, searching for
+    # source definition in `embed/rust/src`
+    source_definition: str = ""  # definition in the source file
+
+    def id(self) -> str:
+        """Identifying this row"""
+        # Used when aggregating alike rows together
+        # Each non-alike row should be always unique, even when no basic info is filled
+        # Need to include a section, not to mix same-named items from different sections
+        def _name_id() -> str:
+            if self.module_name and self.func_name:
+                return f"{self.module_name}::{self.func_name}"
+            elif self.module_name:
+                return self.module_name
+            elif self.func_name:
+                return self.func_name
+            else:
+                return self.symbol_name
+
+        return f"{self.section}_{_name_id()}"
+
+    def format(self, debug: bool = False) -> str:
+        """Nicely formatting this row"""
+        # Definition might not be filled, but when it is, show it instead of the module name
+        # Also, the module name and function name can be empty
+        # Default to symbol name, which is always there
+        if self.source_definition:
+            name_to_show = f"{self.source_definition:<60} {self.func_name}"
+        elif self.module_name:
+            name_to_show = f"{self.module_name:<60} {self.func_name}"
+        elif self.func_name:
+            name_to_show = self.func_name
+        else:
+            name_to_show = self.symbol_name
+
+        to_show = f"{self.section:<10} {self.size:<7_} {self.number_of_symbols:<4} L{self.logic_size:<7_} D{self.data_size:<7_}"
+        if debug:
+            # Optionally showing also the raw symbol name
+            return f"{to_show} {name_to_show:<100} {self.symbol_name}"
+        else:
+            return f"{to_show} {name_to_show}"
+
+
+class RowHandlerAPI(Protocol):
+    """Responsible for handling the row data"""
+
+    language: str
+
+    def add_basic_info(self, row: DataRow) -> DataRow:
+        """Fill in some info, like the module name and the function name."""
+        ...  # pragma: no cover
+
+    def add_definition(self, row: DataRow) -> DataRow:
+        """Include the place where the symbol is defined, if possible."""
+        ...  # pragma: no cover
+
+
+class RowDataLoaderAPI(Protocol):
```

Comment:
I am keeping these `API`s, even though they might seem redundant.

Now when the `binsize` is a released public package, it makes even more sense - it gives an easy way for somebody to write their own extensions/modifications, and supply them instead of "our" default ones. Static type-checkers will then be completely fine with it, because it expects a generic `Protocol`, not some specific class.

Example: somebody has `C++` and `Java` code inside the binary and therefore they will need to implement their own version of `RowHandlerAPI` and pass it into `BinarySize` instead of the default `RowHandlerFactory` (working only with `C`, `upy` and `Rust`). This can be done nicely without any modification of our code - unlike when `BinarySize` (or any other object) would expect a specific object type.